nonnative/src/allocated_nonnative_field_var.rs

```diff
@@ -0,0 +1,982 @@
+            &one,
+            &Boolean::Constant(true),
+        )?;
+        Ok(inverse)
+    }
+
+    /// Convert a `TargetField` element into limbs (not constraints)
+    /// This is an internal function that would be reused by a number of other functions
+    pub fn get_limbs_representations(
+        elem: &TargetField,
+        optimization_type: OptimizationType,
+    ) -> Result<Vec<BaseField>, SynthesisError> {
+        Self::get_limbs_representations_from_big_integer(&elem.into_repr(), optimization_type)
+    }
+
+    /// Obtain the limbs directly from a big int
+    pub fn get_limbs_representations_from_big_integer(
+        elem: &<TargetField as PrimeField>::BigInteger,
+        optimization_type: OptimizationType,
+    ) -> Result<Vec<BaseField>, SynthesisError> {
+        let params = get_params(
+            TargetField::size_in_bits(),
+            BaseField::size_in_bits(),
+            optimization_type,
+        );
+
+        // push the lower limbs first
+        let mut limbs: Vec<BaseField> = Vec::new();
+        let mut cur = *elem;
+        for _ in 0..params.num_limbs {
+            let cur_bits = cur.to_bits_be(); // `to_bits` is big endian
+            let cur_mod_r = <BaseField as PrimeField>::BigInteger::from_bits_be(
+                cur_bits[cur_bits.len() - params.bits_per_limb..].to_vec(),
+            ); // therefore, the lowest `bits_per_non_top_limb` bits is what we want.
+            limbs.push(BaseField::from_repr(cur_mod_r).unwrap());
+            cur.divn(params.bits_per_limb as u32);
+        }
+
+        // then we reserve, so that the limbs are ``big limb first''
+        limbs.reverse();
+
+        Ok(limbs)
+    }
+
+    /// for advanced use, multiply and output the intermediate representations (without reduction)
+    /// This intermediate representations can be added with each other, and they can later be reduced back to the `NonNativeFieldVar`.
+    pub fn mul_without_reduce<CS: ConstraintSystem<BaseField>>(
+        &self,
+        cs: &mut CS,
+        other: &Self,
+    ) -> Result<AllocatedNonNativeFieldMulResultVar<TargetField, BaseField>, SynthesisError> {
+        assert_eq!(self.get_optimization_type(), other.get_optimization_type());
+
+        let params = get_params(
+            TargetField::size_in_bits(),
+            BaseField::size_in_bits(),
+            self.get_optimization_type(),
+        );
+
+        // Step 1: reduce `self` and `other` if necessary
+        let mut self_reduced = self.clone();
+        let mut other_reduced = other.clone();
+        Reducer::<TargetField, BaseField>::pre_mul_reduce(
+            &mut cs.ns(|| "pre_mul_reduce"),
+            &mut self_reduced,
+            &mut other_reduced,
+        )?;
+
+        let mut prod_limbs = Vec::new();
+
+        // We currently only support constraint optimization
+
+        for z_index in 0..2 * params.num_limbs - 1 {
+            prod_limbs.push(FpGadget::alloc_input(
```

Comment:
Should this also be `alloc` instead of `alloc_input`?

Reply:
https://github.com/AleoHQ/snarkVM/pull/78/commits/a4f346c81761a60457a8d5d935a696950f39b6d1